### PR TITLE
Added MesosCluster(String clusterId) constructor for deserializion

### PR DESCRIPTION
**Pull request by frankscholten**
* AbstractContainer - Refactored AbstractContainer.randomId to AbstractContainer.uuid
* AbstractContainer - Added getRole() method and implementation for each subclass
* AbstractContainer - Implemented equals and hashcode based on clusterId, uuid and Docker containerID
* AbstractContainer - Added associateContainerId template method used by subclasses
* Main - Replaced a few static methods with MesosCluster instance methods
* MesosSave - Merged MesosSlaveExtended into MesosSlave
* MesosClusterTest - Added constructor test

### CODE REVIEW

File: minimesos/src/main/java/com/containersol/minimesos/MesosCluster.java [hunk 1]
```diff
@@ -62,6 +62,60 @@ public MesosCluster(ClusterArchitecture clusterArchitecture) {
...
+    public MesosCluster(String clusterId) {
+        this.clusterId = clusterId;
+
+        List<Container> containers = dockerClient.listContainersCmd().exec();
+
+        Collections.sort(containers, (c1,c2) -> Long.compare(c1.getCreated(), c2.getCreated()));
+
+        for (Container container : containers) {
+            for (String name : container.getNames()) {
+                if (name.matches("^/minimesos-.+-" + clusterId + "-\\d+$")) {
```
Comment (sadovnikov) [inline minimesos/src/main/java/com/containersol/minimesos/MesosCluster.java]: To enforce naming convention, should we put composing and parsing containers' names to a separated class?

File: minimesos/src/main/java/com/containersol/minimesos/MesosCluster.java [hunk 2]
```diff
@@ -62,6 +62,60 @@ public MesosCluster(ClusterArchitecture clusterArchitecture) {
...
+                    break;
+                }
+            }
+        }
+
+        if (containers.isEmpty()) {
+            throw new RuntimeException("No containers found for cluster ID " + clusterId);
+        }
+
+        if (getZkContainer() != null) {
```
Comment (sadovnikov) [inline minimesos/src/main/java/com/containersol/minimesos/MesosCluster.java]: ZooKeeper is one of the containers in 'containers'. There is no need to run 'docker ps' one more time

File: minimesos/src/main/java/com/containersol/minimesos/MesosCluster.java [hunk 3]
```diff
@@ -198,7 +310,7 @@ private static void destroyContainers(String clusterId) {
     }
 
     public List<AbstractContainer> getContainers() {
-        return containers;
+        return Collections.unmodifiableList(containers);
```
Comment (sadovnikov) [inline minimesos/src/main/java/com/containersol/minimesos/MesosCluster.java]: Sometimes I also try to protect internal collections, but normally it only complicates things later. Can we return it 'as-is'?

File: minimesos/src/main/java/com/containersol/minimesos/container/AbstractContainer.java [hunk 4]
```diff
@@ -28,21 +28,38 @@
...
+
     private String ipAddress = null;
 
     protected AbstractContainer(DockerClient dockerClient) {
         this.dockerClient = dockerClient;
-        this.randomId = Integer.toUnsignedString(new SecureRandom().nextInt());
+        this.uuid = Integer.toUnsignedString(new SecureRandom().nextInt());
+    }
+
+    protected void associateContainerId() {
```
Comment (sadovnikov) [inline minimesos/src/main/java/com/containersol/minimesos/container/AbstractContainer.java]: This is an expensive method. Please, add missing ID to contractors of all containers. IDs are available in `MesosCluster(String clusterId)`

File: minimesos/src/main/java/com/containersol/minimesos/mesos/ClusterArchitecture.java [hunk 5]
```diff
@@ -174,14 +174,14 @@ public Builder withSlave() {
 
         /**
          * All default instance, but with defined resources
-         * @param resources definition of resources
+         * @param newResources definition of resources
          */
-        public Builder withSlave(String resources) {
+        public Builder withSlave(String newResources) {
```
Comment (sadovnikov) [inline minimesos/src/main/java/com/containersol/minimesos/mesos/ClusterArchitecture.java]: what is the reason for this renaming?

**Comment by sadovnikov** [comment]
@frankscholten I reviewed the PR and put my comments.

Another thing, which I'm happy to be handled separately, if you prefer. I'm glad to see `MesosSlaveExtended` being deleted. Should you consider getting rid of extensions of `MesosMaster`?

image

**Comment by sadovnikov** [comment]
#217 is created for removing extensions of `MesosMaster`

File: minimesos/src/main/java/com/containersol/minimesos/MesosCluster.java [hunk 6]
```diff
@@ -62,6 +62,60 @@ public MesosCluster(ClusterArchitecture clusterArchitecture) {
...
+            container.setClusterId(clusterId);
+        }
+    }
+
+    /**
+     * Recreate a MesosCluster object based on an existing cluster ID.
+     *
+     * @param clusterId the cluster ID of the cluster that is already running
+     */
+    public MesosCluster(String clusterId) {
```
Comment (mwl) [inline minimesos/src/main/java/com/containersol/minimesos/MesosCluster.java]: This _looks_ like something that should have been a factory method instead?

From http://stackoverflow.com/questions/14694982/why-is-using-side-effects-bad-practice-in-javascript-constructors
> Side effects are lies. Your function promises to do one thing, but it also does other hidden things...they are devious and damaging mistruths that often result in strange temporal couplings and order dependencies.

**Comment by sadovnikov** [comment]
I'm implementing the review comments. Will sign when the code is ready for the second round

File: minimesos/src/main/java/com/containersol/minimesos/MesosCluster.java [hunk 7]
```diff
@@ -47,12 +47,13 @@
 
     private static DockerClient dockerClient = DockerClientFactory.build();
```
Comment (frankscholten) [inline minimesos/src/main/java/com/containersol/minimesos/MesosCluster.java]: Does this have to be static?

File: minimesos/src/main/java/com/containersol/minimesos/MesosCluster.java [hunk 8]
```diff
@@ -98,7 +193,34 @@ public void stop() {
...
-     * Start a container. This container will be removed when the Mesos cluster is shut down.
+     * Destroys the Mesos cluster and its containers
+     */
+    public void destroy() {
+        if (clusterId != null) {
+            MarathonClient marathon = new MarathonClient(getMarathonContainer().getIpAddress());
+            marathon.killAllApps();
+
+            DockerClient dockerClient1 = DockerClientFactory.build();
+            List<Container> containers1 = dockerClient1.listContainersCmd().exec();
```
Comment (frankscholten) [inline minimesos/src/main/java/com/containersol/minimesos/MesosCluster.java]: Can you reuse the DockerClient field?

File: minimesos/src/test/java/com/containersol/minimesos/MesosClusterTest.java [hunk 9]
```diff
@@ -43,26 +43,38 @@
     @After
     public void after() {
         DockerContainersUtil util = new DockerContainersUtil(CONFIG.dockerClient);
-        util.getContainers(false).filterByName("^mesos-[0-9a-f\\-]*S\\d*\\.[0-9a-f\\-]*$").remove();
+        util.getContainers(false).filterByName( HelloWorldContainer.CONTAINER_NAME_PATTERN ).kill().remove();
+    }
+
+    @Test
+    public void testConstructor() {
+        String clusterId = CLUSTER.getClusterId();
```
Comment (frankscholten) [inline minimesos/src/test/java/com/containersol/minimesos/MesosClusterTest.java]: Change test name to testLoadCluster

File: minimesos/src/test/java/com/containersol/minimesos/container/ContainerNameTest.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        String[] names = new String[4];
+        names[0] = "/minimesos-agent0/minimesos-zookeeper";
+        names[1] = "/minimesos-agent1/minimesos-zookeeper";
+        names[2] = "/minimesos-agent2/minimesos-zookeeper";
+        names[3] = "/minimesos-zookeeper";
+
+        assertEquals( "minimesos-zookeeper", ContainerName.getFromDockerNames(names));
+
+    }
+
```
Comment (frankscholten) [inline minimesos/src/test/java/com/containersol/minimesos/container/ContainerNameTest.java]: Great!

File: minimesos/src/main/java/com/containersol/minimesos/MesosCluster.java (hunk 7 above)
Comment (sadovnikov) [inline minimesos/src/main/java/com/containersol/minimesos/MesosCluster.java]: Yes, it's used in remaining static methods

**Pull request merged**